File: 08-30/adts/maps/map.cpp
#pragma once

template <class K, class V>
class map
{
public:
  virtual void insert(K key, V value) = 0;
  virtual void remove(K key) = 0;
  virtual bool has(K key) = 0;
  virtual V get(K key) = 0;
  virtual int size() = 0;
  virtual Iterable<K> keys() = 0;
  virtual Iterable<V> values() = 0;
};

File: 08-30/adts/maps/Map.java
package adts.maps;

public interface Map<K, V> {
    public void insert(K key, V value);
    public void remove(K key);
    public boolean has(K key);
    public V get(K key);
    public int size();
    public Iterable<K> keys();
    public Iterable<V> values();
}